sdk/keyvault/Azure.Security.KeyVault.Certificates/src/PemReader.cs

```diff
@@ -0,0 +1,67 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Reflection;
+using System.Security.Cryptography;
+using System.Security.Cryptography.X509Certificates;
+
+namespace Azure.Core
+{
+    /// <summary>
+    /// Reads PEM streams to parse PEM fields or load certificates.
+    /// </summary>
+    internal static partial class PemReader
+    {
+        private static bool s_ecInitializedImportPkcs8PrivateKeyMethod;
+        private static MethodInfo s_ecImportPkcs8PrivateKeyMethod;
+        private static MethodInfo s_ecCopyWithPrivateKeyMethod;
+
+        static partial void CreateECDsaCertificate(byte[] cer, byte[] key, ref X509Certificate2 certificate)
+        {
+            if (!s_ecInitializedImportPkcs8PrivateKeyMethod)
+            {
+                // ImportPkcs8PrivateKey was added in .NET Core 3.0 and is only present on Core. We will fall back to a lightweight decoder if this method is missing from the current runtime.
+                s_ecImportPkcs8PrivateKeyMethod = typeof(ECDsa).GetMethod("ImportPkcs8PrivateKey", BindingFlags.Instance | BindingFlags.Public, null, new[] { typeof(ReadOnlySpan<byte>), typeof(int).MakeByRefType() }, null);
+                s_ecInitializedImportPkcs8PrivateKeyMethod = true;
+            }
+
+            if (s_ecCopyWithPrivateKeyMethod is null)
+            {
+                s_ecCopyWithPrivateKeyMethod = typeof(ECDsaCertificateExtensions).GetMethod("CopyWithPrivateKey", BindingFlags.Static | BindingFlags.Public, null, new[] { typeof(X509Certificate2), typeof(ECDsa) }, null)
+                    ?? throw new PlatformNotSupportedException("The current platform does not support reading a private key from a PEM file");
+            }
+
+            ECDsa privateKey = null;
+            try
+            {
+                if (s_ecImportPkcs8PrivateKeyMethod != null)
+                {
+                    privateKey = ECDsa.Create();
+
+                    // Because ImportECPrivateKeyMethod declares an out parameter we cannot call it directly using MethodInfo.Invoke since all arguments are passed as an object array.
+                    // Instead we create a delegate with the correct signature and invoke it.
+                    ImportPrivateKeyDelegate importECPrivateKey = (ImportPrivateKeyDelegate)s_ecImportPkcs8PrivateKeyMethod.CreateDelegate(typeof(ImportPrivateKeyDelegate), privateKey);
+                    importECPrivateKey.Invoke(key, out _);
+                }
+                else
+                {
+                    privateKey = LightweightPkcs8Decoder.DecodeECDsaPkcs8(key);
+                }
+
+                using X509Certificate2 certificateWithoutPrivateKey = new X509Certificate2(cer);
+
+                // TODO: This will fail for P-256K because the certificate uses OID 1.2.840.10045.1.1 while the private key was modified to use OID 1.3.132.0.10.
```

Comment:
```suggestion
                // TODO: This will fail for P-256K because the certificate uses explicit curve parameters while the private key was modified to use a named curve (1.3.132.0.10).
```

Reply:
You can try something like

```C#
using (ECDsa pubKey = cert.GetECDsaPublicKey())
{
    ECParameters pubParams = pubKey.ExportParameters(false);

    // If the public key uses an explicit curve, force the private key to do the same.
    if (pubParams.Curve.CurveType != ECCurveType.Named)
    {
        ECParameters privExplicit = privateKey.ExportExplicitParameters(true);
        privateKey.ImportParameters(privExplicit);
    }
}
```

Reply:
That would save a lot of risky (or at least of limited scope) manual parsing. I'll see if I can safely refactor the code that way. Using `ECParameters` and related classes (besides `ECDsa`) as you know is not supported <= net47, but in that case I couldn't create an ECDsa certificate anyway. It's just a matter of refactoring the code safely. Since ECDsa keys aren't supported by Azure.Identity anyway (won't work for client auth, of course), I'm trying to limit shared source changes.